[ziiqii] iP

ziiqii opened the pull request:
# UkeCat
```
                    /~(_)~\.          ~ Hello!
 /= ••\      /~(_)~\        \     /\
K=|=|=|=|=|=|=|=|=(  )===]  |    (`~ o7
 \= ••/      \_(~)_/        /   c\   c\
                    \_(~)_/`    U`U_, )=~~  
```

**UkeCat** :octocat: is:
* text-based
* easy to learn
* ~~FAST~~ SUPER FAST to use

All you need to do is:
1. Download it from [here](https://github.com/ziiqii/ip).
2. Double-click to run it!

Features:
- [x] Managing tasks
- [ ] Managing deadlines (coming soon)
- [ ] Reminders (coming soon)

> "*Don't think. Just do.*"

If you are a Java programmer, you can use it to practice Java too. Here's the `updateTasks()`
method from the `FileManager` class:

```java
public static void updateTasks() {
    try (FileWriter fileWriter = new FileWriter("data/taskData.txt")) {
        for (Task task : Storage.getTasks()) {
            fileWriter.write(Parser.parseTaskToCsv(task) + "\n");
        }
    } catch (IOException ioException) {
        System.out.println("IO error occurred while updating.");
    }
}
```

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

Commit:
docs/README.md: Tweak document template

Let's tweak the docs/README.md (which is used as the user guide)
to fit Duke better. Specifically,

1. mention product name in the title
2. mention adding a product screenshot and a product intro
3. tweak the flow to describe feature-by-feature

ziiqii committed:
Implement simple greet and exit messages

ziiqii committed:
Change chatbot name

ziiqii committed:
Change chatbot name, greet and exit msgs

ziiqii committed:
Merge branch 'master' of https://github.com/ziiqii/ip

ziiqii committed:
Add echo to bot

ziiqii committed:
Bot can add words to list, and display list

ziiqii committed:
Use Task object in list

ziiqii committed:
Add ability to mark and unmark tasks

ziiqii committed:
Fix invalid integer input bug

ziiqii committed:
Add subclasses of Task

ziiqii committed:
Add proper replies from bot

ziiqii committed:
Fix unmark bug

ziiqii committed:
Test I/O with redirection

ziiqii committed:
Split main into Parser, Storage, Responder classes. Reduce flavor text for easier testing

ziiqii committed:
Update text testing

ziiqii committed:
Add custom exception class

ziiqii committed:
Add more error handling for marking

ziiqii committed:
Handles more error cases

ziiqii committed:
Already using ArrayList for tasks, standardize call sequence responder -> storage -> parser

ziiqii committed:
Allow deleting tasks

ziiqii committed:
Handle errors for deleting tasks

ziiqii committed:
Add enum for mark type

ziiqii committed:
Update text test

ziiqii committed:
Update switch cases indent

ziiqii committed:
Update switch case indent for mark

ziiqii committed:
Add FileManager class, implement loading data

ziiqii committed:
Implement updateTasks(), rewrite data after bot terminates

ziiqii committed:
Add todo for level-7

ziiqii committed:
Change saved data format to CSV

ziiqii committed:
Implement loading tasks into storage

ziiqii committed:
Overload task constructors to read isDone from CSV

ziiqii committed:
Handle corrupted data file, update data file after every add/delete/mark

ziiqii committed:
Merge branch 'branch-Level-7'

ziiqii committed:
Use LocalDate for by, from to, and fix loading event from data

ziiqii committed:
Print date in different format

ziiqii committed:
Merge branch 'branch-Level-8'

ziiqii committed:
Rename Responder class to Ui, add OOP to UkeCat and Ui

ziiqii committed:
Move ui elements from Storage to Ui, add more OOP

ziiqii committed:
Merge branch 'branch-A-MoreOOP'

ziiqii committed:
Divide classes into packages command, data, enums, exceptions, task

ziiqii committed:
Merge branch 'branch-A-Packages'

ziiqii committed:
Merge gradle support

ziiqii committed:
Update gradle file, change Task to abstract class

ziiqii committed:
Simplify package

ziiqii committed:
Add unit tests for Storage and Parser

ziiqii committed:
Configure build.gradle for jar

ziiqii committed:
Add javadoc comments to all classes

ziiqii committed:
Tweak code to conform to standard

ziiqii committed:
Add ability to find task

ziiqii committed:
Add tweaks

ziiqii committed:
Merge branch 'branch-A-CodingStandard'

ziiqii committed:
Merge branch-A-JavaDoc

ziiqii committed:
Merge branch-Level-9

ziiqii committed:
Add javadoc comments for findTask() and parseFindTask()

ziiqii committed:
Try JavaFX

ziiqii committed:
Try FXML

ziiqii committed:
Update README

ziiqii committed:
Add checkstyle, improve code to comply with style rules

ziiqii committed:
Merge branch 'branch-A-CheckStyle'

ziiqii committed:
Cleanup UkeCat class

AdityaB4 commented on README.md:
````diff
@@ -1,24 +1,39 @@
-# Duke project template
+# UkeCat
+```
+                    /~(_)~\.          ~ Hello!
+ /= ••\      /~(_)~\        \     /\
+K=|=|=|=|=|=|=|=|=(  )===]  |    (`~ o7
+ \= ••/      \_(~)_/        /   c\   c\
+                    \_(~)_/`    U`U_, )=~~  
+```
````
I love how nice this looks and matches with your theme

AdityaB4 commented on src/main/java/ukecat/Parser.java:
```diff
@@ -0,0 +1,214 @@
+    /**
+     * Parses a CSV (Comma-Separated Values) format string and converts it into a format
+     * compatible with the UkeCat application. The CSV format is expected to represent tasks
+     * in the following way:
+     * - ToDo: T, 0/1, desc
+     * - Deadline: D, 0/1, desc, by
+     * - Event: E, 0/1, desc, from, to
+     * The method extracts information from the CSV string, modifies the content in the Storage
+     * class, and then calls the respective parsing methods for ToDo, Deadline, or Event tasks.
+     *
+     * @param csv The CSV format string to be parsed.
+     */
+    public static void parseCsv(String csv) {
+        String[] words = csv.split(",");
+        Storage.words = words;
+        try {
+            switch (words[0]) {
+            case "T":
+                Storage.words[0] = "todo";
+                Storage.input = String.format("todo %s", words[2]);
+                parseToDo(Storage.input);
+                break;
+            case "D":
+                Storage.words[0] = "deadline";
+                Storage.input = String.format("deadline %s /by %s", words[2], words[3]);
+                parseDeadline(Storage.input);
+                break;
+            case "E":
+                Storage.words[0] = "event";
+                Storage.input = String.format("event %s /from %s /to %s", words[2], words[3], words[4]);
+                parseEvent(Storage.input);
+                break;
+            default:
+                System.out.println("Can't read csv, file corrupted, abort!");
+                System.exit(-1);
+            }
+        } catch (UkeCatException e) {
+            System.out.println(e.getMessage());
+        }
+    }
```
Parsing to and from csv is a great idea!

AdityaB4 commented on src/main/java/ukecat/Storage.java:
```diff
@@ -0,0 +1,207 @@
+ * It includes methods for adding tasks and accessing the list of tasks.
+ */
+public class Storage {
+    // ukecat.Storage from input
+    public static String input; // user input
+    public static String[] words; // split words using
+    public static String desc; // for any task
```
Do you think the comments here might be a little redundant? 🤔

AdityaB4 commented on src/main/java/ukecat/Task.java:
```diff
@@ -0,0 +1,87 @@
+        System.out.println("    " + this);
+    }
+
+    /**
+     * Gets the integer representation of the isDone status (1 for done, 0 for not done).
+     *
+     * @return The integer representation of the isDone status.
+     */
+    public int getIntIsDone() {
+        return isDone ? 1 : 0;
+    }
```
I don't like the usage of integers here. Consider using an enum, so that adding a new status (like "in progress") would be cleaner and more readable.

AdityaB4 reviewed (commented):
Clean code that complies with the standard, with a lot of detailed text-ui customisations! LGTM

rahhulleee commented on src/main/java/ukecat/Deadline.java:
```diff
@@ -0,0 +1,56 @@
+
+    /**
+     * Gets the string representation of the deadline date.
+     *
```
i like the documentation of the class and methods.

rahhulleee commented on src/main/java/ukecat/Event.java:
```diff
@@ -0,0 +1,72 @@
+     *
+     * @return The start date of the event as a string.
+     */
+    public String getStart() {
```
clear and good naming of the methods in classes. reader can easily understand what the getStart() method does and is not ambiguous

rahhulleee commented on src/main/java/ukecat/Parser.java:
```diff
@@ -0,0 +1,214 @@
+     * @param t The Task object to be converted to CSV.
+     * @return A CSV format string representing the given Task.
+     */
+    public static String parseTaskToCsv(Task t) {
```
once again, methods are relevantly named. good job

rahhulleee commented on src/main/java/ukecat/Storage.java:
```diff
@@ -0,0 +1,207 @@
+    public static LocalDate start; // for events
+    public static LocalDate end; // for events
+
+    public static int numT = 0;
```
numT could be changed to a better and more representative name. perhaps numOfTasks?

rahhulleee reviewed (commented):
most of the methods have names that tells reader , in a clear cut manner, what the method is created for. so good job !

ziiqii committed:
Link original UkeCat functions to GUI

ziiqii committed:
Stop tracking data/taskData.txt

ziiqii committed:
Modify welcome msg in GUI

ziiqii committed:
Add waiting time before termination with 'bye'

ziiqii committed:
Adopt TaskStatus enum

ziiqii committed:
Update GUI

ziiqii committed:
Merge branch 'branch-Level-10'

ziiqii committed:
Fix gradle file to create .jar

ziiqii committed:
Add assertions

Assertions help to define assumptions about the program.

In the MainWindow class, assertions ensure the nodes are
initialized properly for the GUI to display.

In the Storage class, assertions verify that user inputs are stored
correctly for further usage by other methods.

Let's add assertions into these classes to make the code more robust.

ziiqii committed:
Improve code quality

Better code quality improves readability and robustness of code.

Let's
* replace magic numbers with static final variables
* use guard clauses to make the happy path prominent
* ensure default case exists for every switch statement
* delete dead code

ziiqii committed:
Use Java Streams to find tasks

Streams offer improved code readability and conciseness. They also
support parallelism, which could lead to improved performance
improvements for the future.

So let's use streams to replace some manual iteration loops.

ziiqii committed:
Merge pull request #2 from ziiqii/branch-A-Assertions

Add assertions

ziiqii committed:
Merge branch 'master' of https://github.com/ziiqii/ip

ziiqii committed:
Merge master to branch-A-CodeQuality

ziiqii committed:
Merge branch 'master' into branch-A-Streams

ziiqii committed:
Merge pull request #3 from ziiqii/branch-A-CodeQuality

Branch a code quality

ziiqii committed:
Merge branch 'master' of https://github.com/ziiqii/ip

ziiqii committed:
Merge branch 'master' into branch-A-Streams

ziiqii committed:
Merge pull request #4 from ziiqii/branch-A-Streams

Branch a streams

ziiqii committed:
Merge branch 'master' of https://github.com/ziiqii/ip

ziiqii committed:
Add yml workflow file

ziiqii committed:
Improve codestyle

ziiqii committed:
Update JUnit tests

ziiqii committed:
Fix style

ziiqii committed:
Update yaml file

ziiqii committed:
Add ability to handle recurring tasks

ziiqii committed:
Add javadoc for recurring task methods

ziiqii committed:
Merge pull request #5 from ziiqii/branch-B-RecurringTasks

Branch b recurring tasks

ziiqii committed:
Merge branch 'master' of https://github.com/ziiqii/ip

ziiqii committed:
Fix file updates

ziiqii committed:
Clean up code

ziiqii committed:
Add user guide and update readme

ziiqii committed:
Move Ui.png

ziiqii committed:
Update user guide image

ziiqii committed:
Update user guide

ziiqii committed:
Add personality to responses

ziiqii committed:
Merge branch 'branch-A-Personality'

ziiqii committed:
Upgrade GUI colors

ziiqii committed:
Merge branch 'branch-A-BetterGui'

ziiqii committed:
Update UI png

ziiqii committed:
Add more testing

ziiqii committed:
Handle event date error

ziiqii committed:
Prevent resizing GUI

ziiqii committed:
Fix bug

ziiqii committed:
Compress images

ziiqii committed:
Fix typo
